### PR TITLE
"Too many open files" errors cause some targets

### PR DESCRIPTION
**Pull request by lodos2005**
- 5936936406ebf10516f2d0e95a6a72a740cfa91a : #489 increasing ulimit -n and ulimit -u, along with this fix, prevented me from encountering the mentioned error.

### CODE REVIEW

File: cme/protocols/smb.py [hunk 1]
```diff
@@ -607,7 +607,8 @@ def create_smbv3_conn(self, kdc=""):
             self.smbv1 = False
         except socket.error as e:
             if str(e).find("Too many open files") != -1:
-                self.logger.fail(f"SMBv3 connection error on {self.host if not kdc else kdc}: {e}")
+                if not logger is None:
```
Comment (mpgn) [inline cme/protocols/smb.py:610]: logger ?

Comment (lodos2005) [reply cme/protocols/smb.py:610]: Yep there is stacktrace

```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/lodos2005/GIT/CrackMapExec/cme/protocols/smb.py", line 571, in create_smbv3_conn
self.conn = SMBConnection(self.host if not kdc else kdc, self.host if not kdc else kdc, None, self.args.port, timeout=self.args.smb_timeout)
File "/home/lodos2005/.cache/pypoetry/virtualenvs/crackmapexec-MRXVD3-M-py3.10/lib/python3.10/site-packages/impacket/smbconnection.py", line 80, in init
self.negotiateSession(preferredDialect)
File "/home/lodos2005/.cache/pypoetry/virtualenvs/crackmapexec-MRXVD3-M-py3.10/lib/python3.10/site-packages/impacket/smbconnection.py", line 120, in negotiateSession
packet = self.negotiateSessionWildcard(self._myName, self._remoteName, self._remoteHost, self._sess_port,
File "/home/lodos2005/.cache/pypoetry/virtualenvs/crackmapexec-MRXVD3-M-py3.10/lib/python3.10/site-packages/impacket/smbconnection.py", line 169, in negotiateSessionWildcard
self._nmbSession = nmb.NetBIOSTCPSession(myName, remoteName, remoteHost, nmb.TYPE_SERVER, sess_port,
File "/home/lodos2005/.cache/pypoetry/virtualenvs/crackmapexec-MRXVD3-M-py3.10/lib/python3.10/site-packages/impacket/nmb.py", line 893, in init
NetBIOSSession.init(self, myname, remote_name, remote_host, remote_type=remote_type, sess_port=sess_port,
File "/home/lodos2005/.cache/pypoetry/virtualenvs/crackmapexec-MRXVD3-M-py3.10/lib/python3.10/site-packages/impacket/nmb.py", line 753, in init
self._sock = self._setup_connection((remote_host, sess_port), timeout)
File "/home/lodos2005/.cache/pypoetry/virtualenvs/crackmapexec-MRXVD3-M-py3.10/lib/python3.10/site-packages/impacket/nmb.py", line 905, in _setup_connection
raise socket.error("Connection error (%s:%s)" % (peer[0], peer[1]), e)
OSError: [Errno Connection error (5.136.20.14:445)] [Errno 24] Too many open files

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "", line 1, in
File "/home/lodos2005/GIT/CrackMapExec/cme/crackmapexec.py", line 257, in main
asyncio.run(
File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
return loop.run_until_complete(main)
File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
return future.result()
File "/home/lodos2005/GIT/CrackMapExec/cme/crackmapexec.py", line 105, in start_threadpool
await asyncio.gather(*jobs)
File "/home/lodos2005/GIT/CrackMapExec/cme/crackmapexec.py", line 69, in run_protocol
await asyncio.wait_for(
File "/usr/lib/python3.10/asyncio/tasks.py", line 408, in wait_for
return await fut
File "/usr/lib/python3.10/concurrent/futures/thread.py", line 58, in run
result = self.fn(*self.args, **self.kwargs)
File "/home/lodos2005/GIT/CrackMapExec/cme/protocols/smb.py", line 143, in init
connection.init(self, args, db, host)
File "/home/lodos2005/GIT/CrackMapExec/cme/connection.py", line 65, in init
self.proto_flow()
File "/home/lodos2005/GIT/CrackMapExec/cme/connection.py", line 96, in proto_flow
if self.create_conn_obj():
File "/home/lodos2005/GIT/CrackMapExec/cme/protocols/smb.py", line 585, in create_conn_obj
elif self.create_smbv3_conn(kdc):
File "/home/lodos2005/GIT/CrackMapExec/cme/protocols/smb.py", line 574, in create_smbv3_conn
self.logger.error('SMBv3 connection error on {}: {}'.format(self.host if not kdc else kdc, e))
AttributeError: 'NoneType' object has no attribute 'error'
```


Sometimes the logger is not being set and it doesn't output an error message. As a result, the program crashes.

```
self.logger.error('SMBv3 connection error on {}: {}'.format(self.host if not kdc else kdc, e))
AttributeError: 'NoneType' object has no attribute 'error'
```

Comment (mpgn) [reply cme/protocols/smb.py:610]: i think we can go if not self.logger: print()

print will never fail on us aha

Comment (lodos2005) [reply cme/protocols/smb.py:610]: i think it relevent file descriptors but idk why. I have only encountered the issue of the logger not being set in this line.  If the logger is NoneType, perhaps we can use print instead like you said.

Comment (NeffIsBack) [reply cme/protocols/smb.py:610]: Is this reproducable? Can you trigger the error somehow? Would be interesting to `print(self.logger)` after the initial [proto_logger](https://github.com/mpgn/CrackMapExec/blob/ab692c5b9e25c6301a82d1570825099d33b3a384/cme/protocols/smb.py#L173) function. And what the value is and why it isn't setting the logger before creating the connection.

If it's even executing that function before it somehow somewhere tries to create an connection object

Comment (NeffIsBack) [reply cme/protocols/smb.py:610]: > Is this reproducable? Can you trigger the error somehow? Would be interesting to `print(self.logger)` after the initial [proto_logger](https://github.com/mpgn/CrackMapExec/blob/ab692c5b9e25c6301a82d1570825099d33b3a384/cme/protocols/smb.py#L173) function. And what the value is and why it isn't setting the logger before creating the connection.
> 
> If it's even executing that function before it somehow somewhere tries to create an connection object

Could you perhaps also print out what `self` is in [this](https://github.com/mpgn/CrackMapExec/blob/ab692c5b9e25c6301a82d1570825099d33b3a384/cme/protocols/smb.py#L610) line? Maybe this is some other arbitrary object whyever this is not be the SMB protocol class

Comment (lodos2005) [reply cme/protocols/smb.py:610]: > Is this reproducable? Can you trigger the error somehow? Would be interesting to `print(self.logger)` after the initial [proto_logger](https://github.com/mpgn/CrackMapExec/blob/ab692c5b9e25c6301a82d1570825099d33b3a384/cme/protocols/smb.py#L173) function. And what the value is and why it isn't setting the logger before creating the connection.
> 
> If it's even executing that function before it somehow somewhere tries to create an connection object

I don't currently have an environment that can generate this error, but it's a problem that occurs when we target multiple hosts.

Comment (mpgn) [reply cme/protocols/smb.py:610]: @NeffIsBack set the logger to None just before the smbv3 connection, just force it :)

Comment (NeffIsBack) [reply cme/protocols/smb.py:610]: How about we do
```python
if not self.logger:
   print("DEBUG the logger error, please open an issue on github: " + str(self) + str(self.logger))
   self.proto_logger()
self.logger.fail...
```

If that does not solve the problem `self` is most likely not the smb protocol

Comment (mpgn) [reply cme/protocols/smb.py:610]: propose the change in a PR, it will be easy to review :)

Comment (NeffIsBack) [reply cme/protocols/smb.py:610]: @lodos2005 do you still have the vm/machine you encountered the issue on? Can you give me the output of `$ ulimit -a`?

Comment (lodos2005) [reply cme/protocols/smb.py:610]: > @lodos2005 do you still have the vm/machine you encountered the issue on? Can you give me the output of `ulimit -a` and your OS?

`ulimit -a`
```
-t: cpu time (seconds)              unlimited
-f: file size (blocks)              unlimited
-d: data seg size (kbytes)          unlimited
-s: stack size (kbytes)             8192
-c: core file size (blocks)         0
-m: resident set size (kbytes)      unlimited
-u: processes                       11537
-n: file descriptors                1024
-l: locked-in-memory size (kbytes)  381172
-v: address space (kbytes)          unlimited
-x: file locks                      unlimited
-i: pending signals                 11537
-q: bytes in POSIX msg queues       819200
-e: max nice                        0
-r: max rt priority                 0
-N 15: rt cpu time (microseconds)   unlimited
```


`hostnamectl`
```
Static hostname: kali
       Icon name: computer-vm
         Chassis: vm 🖴
      Machine ID: b0efcab8b3084211be1ab7d22ae6fc27
         Boot ID: b8cb215aaf8845ad9d39b7be33c3ff38
  Virtualization: vmware
Operating System: Kali GNU/Linux Rolling          
          Kernel: Linux 5.18.0-kali5-amd64
    Architecture: x86-64
 Hardware Vendor: VMware, Inc.
  Hardware Model: VMware Virtual Platform
Firmware Version: 6.00
```

`lsb_release -a`

```
Distributor ID: Kali
Description:    Kali GNU/Linux Rolling
Release:        2022.3
Codename:       kali-rolling
```

**Comment by mpgn** [comment]
closed in favor of #27

**Pull request closed without merging**